File: py-le/merkle_tree.py
"""Merkle Tree implementation in Python"""

import hashlib
from typing import List, Optional, Tuple


class MerkleNode:
    """Node in a Merkle tree"""
    
    def __init__(self, hash_value: str, data: Optional[str] = None, left: Optional['MerkleNode'] = None, right: Optional['MerkleNode'] = None):
        self.hash = hash_value
        self.data = data
        self.left = left
        self.right = right
        self.is_leaf = data is not None
    
    def __repr__(self) -> str:
        return f"MerkleNode(hash={self.hash[:8]}..., is_leaf={self.is_leaf})"


class MerkleTree:
    """Binary Merkle Tree implementation"""
    
    def __init__(self, data_blocks: Optional[List[str]] = None, hash_func: str = "sha256"):
        """
        Initialize Merkle tree
        
        Args:
            data_blocks: List of data strings to hash
            hash_func: Hash function to use ('sha256', 'sha1', 'md5')
        """
        self.hash_func = hash_func
        self.root: Optional[MerkleNode] = None
        self.leaves: List[MerkleNode] = []
        self.tree_nodes: List[MerkleNode] = []
        
        if data_blocks:
            self.build(data_blocks)
    
    @staticmethod
    def _hash(data: str, hash_func: str = "sha256") -> str:
        """Compute hash of data"""
        if hash_func == "sha256":
            return hashlib.sha256(data.encode()).hexdigest()
        elif hash_func == "sha1":
            return hashlib.sha1(data.encode()).hexdigest()
        elif hash_func == "md5":
            return hashlib.md5(data.encode()).hexdigest()
        else:
            raise ValueError(f"Unknown hash function: {hash_func}")
    
    def build(self, data_blocks: List[str]) -> None:
        """Build Merkle tree from data blocks"""
        if not data_blocks:
            raise ValueError("Data blocks cannot be empty")
        
        # Create leaf nodes
        self.leaves = []
        for data in data_blocks:
            hash_value = self._hash(data, self.hash_func)
            leaf = MerkleNode(hash_value, data=data)
            self.leaves.append(leaf)
            self.tree_nodes.append(leaf)
        
        # Build tree bottom-up
        current_level = self.leaves[:]
        
        while len(current_level) > 1:
            next_level = []
            
            # Process pairs
            for i in range(0, len(current_level), 2):
                left = current_level[i]
                right = current_level[i + 1] if i + 1 < len(current_level) else left
                
                # Combine hashes
                combined = left.hash + right.hash
                parent_hash = self._hash(combined, self.hash_func)
                parent = MerkleNode(parent_hash, left=left, right=right)
                
                next_level.append(parent)
                self.tree_nodes.append(parent)
            
            current_level = next_level
        
        # Set root
        self.root = current_level[0] if current_level else None
    
    def get_root_hash(self) -> str:
        """Get the root hash of the tree"""
        if not self.root:
            raise RuntimeError("Tree is not built")
        return self.root.hash
    
    def get_proof(self, leaf_index: int) -> List[Tuple[str, str]]:
        """
        Get Merkle proof for a leaf node
        Returns list of (sibling_hash, position) tuples where position is 'left' or 'right'
        """
        if leaf_index >= len(self.leaves):
            raise IndexError(f"Leaf index {leaf_index} out of range")
        
        proof = []
        current = self.leaves[leaf_index]
        
        # Find path from leaf to root
        def find_proof_path(node: MerkleNode, target: MerkleNode, path: List[MerkleNode]) -> bool:
            if node is target:
                return True
            
            if node.left is None:
                return False
            
            if find_proof_path(node.left, target, path):
                # Add right sibling
                if node.right:
                    proof.append((node.right.hash, "left"))
                return True
            
            if node.right and find_proof_path(node.right, target, path):
                # Add left sibling
                proof.append((node.left.hash, "right"))
                return True
            
            return False
        
        find_proof_path(self.root, current, [])
        return proof
    
    def verify_leaf(self, leaf_index: int, data: str, proof: List[Tuple[str, str]]) -> bool:
        """
        Verify a leaf using Merkle proof
        
        Args:
            leaf_index: Index of the leaf
            data: Data that should hash to leaf
            proof: Merkle proof (list of sibling hashes)
        
        Returns:
            True if leaf is valid
        """
        # Compute leaf hash
        current_hash = self._hash(data, self.hash_func)
        
        # Traverse proof path
        for sibling_hash, position in proof:
            if position == "left":
                current_hash = self._hash(current_hash + sibling_hash, self.hash_func)
            else:
                current_hash = self._hash(sibling_hash + current_hash, self.hash_func)
        
        return current_hash == self.get_root_hash()
    
    def get_height(self) -> int:
        """Get height of the tree"""
        if not self.root:
            return 0
        
        def height(node: Optional[MerkleNode]) -> int:
            if node is None or node.is_leaf:
                return 1
            return 1 + max(height(node.left), height(node.right))
        
        return height(self.root)
    
    def display(self) -> str:
        """Display tree structure"""
        result = []
        result.append(f"Merkle Tree (hash={self.hash_func})")
        result.append(f"Number of leaves: {len(self.leaves)}")
        result.append(f"Height: {self.get_height()}")
        result.append(f"Root hash: {self.get_root_hash()}")
        
        result.append("\nLeaves:")
        for i, leaf in enumerate(self.leaves):
            result.append(f"  [{i}] {leaf.data} -> {leaf.hash[:16]}...")
        
        return "\n".join(result)
    
    def visualize_tree(self, node: Optional[MerkleNode] = None, prefix: str = "", is_tail: bool = True) -> str:
        """Visualize tree structure"""
        if node is None:
            node = self.root
            if not node:
                return "Empty tree"
        
        result = []
        
        # Add current node
        connector = "└── " if is_tail else "├── "
        label = f"{node.data} [{node.hash[:8]}...]" if node.is_leaf else f"[{node.hash[:8]}...]"
        result.append(prefix + connector + label)
        
        # Add children
        if node.left or node.right:
            extension = "    " if is_tail else "│   "
            
            if node.left:
                child_result = self.visualize_tree(node.left, prefix + extension, node.right is None)
                result.append(child_result)
            
            if node.right:
                child_result = self.visualize_tree(node.right, prefix + extension, True)
                result.append(child_result)
        
        return "\n".join(result)

File: go-le/merkle_tree.go
package main

import (
	"crypto/sha256"
	"fmt"
	"strings"
)

// MerkleNode represents a node in the Merkle tree
type MerkleNode struct {
	Hash   string
	Data   string
	Left   *MerkleNode
	Right  *MerkleNode
	IsLeaf bool
}

// MerkleProof represents a proof for a leaf verification
type MerkleProof struct {
	SiblingHash string
	Position    string // "left" or "right"
}

// MerkleTree represents a binary Merkle tree
type MerkleTree struct {
	Root     *MerkleNode
	Leaves   []*MerkleNode
	HashFunc string
}

// NewMerkleTree creates a new Merkle tree from data blocks
func NewMerkleTree(dataBlocks []string, hashFunc string) (*MerkleTree, error) {
	if len(dataBlocks) == 0 {
		return nil, fmt.Errorf("data blocks cannot be empty")
	}

	tree := &MerkleTree{
		HashFunc: hashFunc,
	}

	tree.Build(dataBlocks)
	return tree, nil
}

// Hash computes the hash of data
func (mt *MerkleTree) Hash(data string) string {
	if mt.HashFunc != "sha256" {
		mt.HashFunc = "sha256"
	}
	hash := sha256.Sum256([]byte(data))
	return fmt.Sprintf("%x", hash)
}

// Build constructs the Merkle tree from data blocks
func (mt *MerkleTree) Build(dataBlocks []string) {
	// Create leaf nodes
	mt.Leaves = make([]*MerkleNode, len(dataBlocks))
	for i, data := range dataBlocks {
		hash := mt.Hash(data)
		leaf := &MerkleNode{
			Hash:   hash,
			Data:   data,
			IsLeaf: true,
		}
		mt.Leaves[i] = leaf
	}

	// Build tree bottom-up
	currentLevel := make([]*MerkleNode, len(mt.Leaves))
	copy(currentLevel, mt.Leaves)

	for len(currentLevel) > 1 {
		var nextLevel []*MerkleNode

		// Process pairs
		for i := 0; i < len(currentLevel); i += 2 {
			left := currentLevel[i]
			right := currentLevel[i]

			// Handle odd number of nodes
			if i+1 < len(currentLevel) {
				right = currentLevel[i+1]
			}

			// Combine hashes
			combined := left.Hash + right.Hash
			parentHash := mt.Hash(combined)
			parent := &MerkleNode{
				Hash:   parentHash,
				Left:   left,
				Right:  right,
				IsLeaf: false,
			}

			nextLevel = append(nextLevel, parent)
		}

		currentLevel = nextLevel
	}

	// Set root
	if len(currentLevel) > 0 {
		mt.Root = currentLevel[0]
	}
}

// GetRootHash returns the root hash of the tree
func (mt *MerkleTree) GetRootHash() string {
	if mt.Root == nil {
		return ""
	}
	return mt.Root.Hash
}

// GetProof generates a Merkle proof for a leaf at given index
func (mt *MerkleTree) GetProof(leafIndex int) []MerkleProof {
	if leafIndex >= len(mt.Leaves) {
		return nil
	}

	var proof []MerkleProof
	target := mt.Leaves[leafIndex]
	mt.findProofPath(mt.Root, target, &proof)

	// Reverse proof to go from leaf to root
	for i, j := 0, len(proof)-1; i < j; i, j = i+1, j-1 {
		proof[i], proof[j] = proof[j], proof[i]
	}

	return proof
}

// findProofPath recursively finds the proof path for a leaf
func (mt *MerkleTree) findProofPath(node *MerkleNode, target *MerkleNode, proof *[]MerkleProof) bool {
	if node == nil {
		return false
	}

	if node == target {
		return true
	}

	if node.Left == nil {
		return false
	}

	// Search left subtree
	if mt.findProofPath(node.Left, target, proof) {
		// Add right sibling
		if node.Right != nil {
			*proof = append([]MerkleProof{{SiblingHash: node.Right.Hash, Position: "right"}}, (*proof)...)
		}
		return true
	}

	// Search right subtree
	if node.Right != nil && mt.findProofPath(node.Right, target, proof) {
		// Add left sibling
		*proof = append([]MerkleProof{{SiblingHash: node.Left.Hash, Position: "left"}}, (*proof)...)
		return true
	}

	return false
}

// VerifyLeaf verifies a leaf using Merkle proof
func (mt *MerkleTree) VerifyLeaf(leafIndex int, data string, proof []MerkleProof) bool {
	// Compute leaf hash
	currentHash := mt.Hash(data)

	// Traverse proof path
	for _, p := range proof {
		if p.Position == "left" {
			currentHash = mt.Hash(p.SiblingHash + currentHash)
		} else {
			currentHash = mt.Hash(currentHash + p.SiblingHash)
		}
	}

	return currentHash == mt.GetRootHash()
}

// GetHeight returns the height of the tree
func (mt *MerkleTree) GetHeight() int {
	if mt.Root == nil {
		return 0
	}
	return mt.getNodeHeight(mt.Root)
}

// getNodeHeight recursively computes node height
func (mt *MerkleTree) getNodeHeight(node *MerkleNode) int {
	if node == nil || node.IsLeaf {
		return 1
	}
	return 1 + max(mt.getNodeHeight(node.Left), mt.getNodeHeight(node.Right))
}

// Display returns a string representation of the tree
func (mt *MerkleTree) Display() string {
	result := []string{
		fmt.Sprintf("Merkle Tree (hash=%s)", mt.HashFunc),
		fmt.Sprintf("Number of leaves: %d", len(mt.Leaves)),
		fmt.Sprintf("Height: %d", mt.GetHeight()),
		fmt.Sprintf("Root hash: %s", mt.GetRootHash()),
		"",
		"Leaves:",
	}

	for i, leaf := range mt.Leaves {
		hash := leaf.Hash
		if len(hash) > 16 {
			hash = hash[:16] + "..."
		}
		result = append(result, fmt.Sprintf("  [%d] %s -> %s", i, leaf.Data, hash))
	}

	return strings.Join(result, "\n")
}

// VisualizeTree returns an ASCII visualization of the tree
func (mt *MerkleTree) VisualizeTree() string {
	if mt.Root == nil {
		return "Empty tree"
	}
	return mt.visualizeNode(mt.Root, "", true)
}

// visualizeNode recursively visualizes tree nodes
func (mt *MerkleTree) visualizeNode(node *MerkleNode, prefix string, isTail bool) string {
	if node == nil {
		return ""
	}

	connector := "└── "
	if !isTail {
		connector = "├── "
	}

	hash := node.Hash
	if len(hash) > 8 {
		hash = hash[:8] + "..."
	}

	var label string
	if node.IsLeaf {
		label = fmt.Sprintf("%s [%s]", node.Data, hash)
	} else {
		label = fmt.Sprintf("[%s]", hash)
	}

	result := prefix + connector + label + "\n"

	if node.Left != nil || node.Right != nil {
		extension := "    "
		if !isTail {
			extension = "│   "
		}

		if node.Left != nil {
			result += mt.visualizeNode(node.Left, prefix+extension, node.Right == nil)
		}

		if node.Right != nil {
			result += mt.visualizeNode(node.Right, prefix+extension, true)
		}
	}

	return result
}

// Helper function
func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}
